app/src/main/java/com/github/epfl/meili/forum/ForumActivity.kt

```diff
@@ -0,0 +1,143 @@
+package com.github.epfl.meili.forum
+
+import android.content.Intent
+import android.graphics.Color
+import android.os.Bundle
+import android.view.View
+import android.widget.LinearLayout
+import android.widget.TextView
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.Observer
+import com.github.epfl.meili.R
+import com.github.epfl.meili.home.Auth
+import com.github.epfl.meili.home.GoogleSignInActivity
+
+
+class ForumActivity : AppCompatActivity() {
+
+    private val TAG = "ForumActivity"
+    private var shownPosts = ArrayList<Post>()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_forum)
+
+        // Create observer that makes a UI for each post in the observed list
+        val forumObserver = Observer<List<Post>> { posts ->
+            updatePosts(posts) // Update the posts in UI
+        }
+        // Observe the posts from viewModel
+        ForumViewModel.posts.observe(this, forumObserver)
+    }
+
+    /** Called when the user taps a post */
+    fun openPost(view: View, post_id: String) {
+        val intent = Intent(this, PostActivity::class.java).apply {
+            putExtra(
+                EXTRA_POST_ID,
+                post_id
+            ) // pass ID to PostActivity so it knows which one to fetch
+        }
+        startActivity(intent) // starts the instance of PostActivity
+    }
+
+    /** Called when the user taps the + button */
+    fun goToPostCreation(view: View) {
+        // Only create post if logged in, Otherwise ask to log in first
+        val intent: Intent =
+            if (Auth.name != null) {
+                Intent(this, NewPostActivity::class.java)
+            } else {
+                Intent(this, GoogleSignInActivity::class.java)
+            }
+
+        startActivity(intent) // starts the instance
+    }
+
+    /** Update the forum from new database information */
+    private fun updatePosts(posts: List<Post>) {
+        val newPosts = posts.minus(shownPosts)
+        for (post in newPosts) {
+            addPostToForum(post) // add new Post to forum
+        }
+    }
+
+    /** Add a new post to the forum UI */
+    private fun addPostToForum(post: Post) {
+        val layout = makePostBox(post.id)
+        addAuthor(post.author, layout)
+        addTitle(post.title, layout)
+        shownPosts.add(post)
+    }
+
+    /** Makes the clickable box for the post information to go into */
+    private fun makePostBox(post_id: String): LinearLayout {
+        // Create vertical linear layout (box)
+        val box = LinearLayout(this)
```

Comment:
Since I'm implementing an upvote/downvote system soon, can I know what API you are using for the forum boxes? I think it will be helpful for me.

Reply:
I use a LinearLayout that stores the Title, Text and author as TextViews. And then I use one LinearLayout that contains those linear layouts.

There's probably other ways of doing it, but that was the most convenient for me at the moment.

Reply:
Thank you!